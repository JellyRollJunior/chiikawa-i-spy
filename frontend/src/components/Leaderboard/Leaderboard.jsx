import { useEffect, useState } from 'react';
import { useWinners } from '../../hooks/useWinners';
import { IconWrapper } from '../IconWrapper/IconWrapper.jsx';
import styles from './Leaderboard.module.css';
import shared from '../../styles/shared.module.css';

const Leaderboard = ({ gameIds }) => {
  const { winners, loading, error } = useWinners();
  const [gameId, setGameId] = useState(null);

  useEffect(() => {
    gameIds && setGameId(gameIds[0])
  }, [gameIds])

  return (
    <section className={`${shared.marginTopMedium}`}>
      <IconWrapper size={46} margin={12}>
        <h2 className={`${shared.card} ${shared.title}`}>Leaderboard</h2>
      </IconWrapper>
      {loading && <h3 className={shared.messages}>Loading winners</h3>}
      {error && <h3 className={shared.messages}>{error}</h3>}
      <ul className={`${styles.leaderboard} ${shared.card} ${shared.marginTopSmall}`}>
        <li className={styles.leaderboardLegend} key={'he was number one!'}>
          <h3>Place</h3>
          <h3>Name</h3>
          <h3>Time</h3>
        </li>
        {winners && gameId && winners.map(
          (winner, index) =>
            winner.gameId == gameId && 
              <li className={styles.leaderboardItem} key={winner.id}>
                <h4>{index}</h4>
                <h4>{winner.name}</h4>
                <h4 className={styles.left}>{winner.seconds}s</h4>
              </li> 
          )}
      </ul>
    </section>
  );
};

export { Leaderboard };
