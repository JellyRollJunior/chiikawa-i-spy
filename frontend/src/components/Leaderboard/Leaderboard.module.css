.leaderboard {
  display: flex;
  flex-direction: column;

  min-width: 300px;
  max-width: 450px;
  width: 100%;
  padding: 16px 12px;
}

.leaderboardItem {
  display: grid;
  grid-template-columns: 50px 1fr 80px;
}

.leaderboardItem * {
  text-align: center;
}
