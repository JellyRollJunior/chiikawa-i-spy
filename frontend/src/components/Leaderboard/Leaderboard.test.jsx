import { vi, describe, it, expect } from 'vitest';
import { screen, render } from '@testing-library/react';
import { userEvent } from '@testing-library/user-event';
import { Leaderboard } from './Leaderboard.jsx';

const GAME_2_NAME = 'game 2';
const GAME_2_PLAYER = 'usagi';
const GAME_DATA = [
  {
    id: 1,
    name: 'game 1',
  },
  {
    id: 2,
    name: GAME_2_NAME,
  },
];
const WINNER_DATA = [
  {
    gameId: 1,
    id: 30,
    name: 'Anonymous mob character',
    seconds: '7.383',
  },
  {
    gameId: 2,
    id: 13,
    name: GAME_2_PLAYER,
    seconds: '11.327',
  },
];

vi.mock('../../hooks/useWinners.js', () => ({
  useWinners: () => {
    return { winners: WINNER_DATA };
  },
}));

vi.mock('../IconWrapper/IconWrapper.jsx', () => ({
  IconWrapper: ({ children }) => children,
}));

describe('Leaderboard component', () => {
  it('renders title, game buttons, and winners', () => {
    const { container } = render(<Leaderboard games={GAME_DATA} />);

    expect(container).toMatchSnapshot();
  });

  it('renders winner from game 2 when game 2 button is clicked', async () => {
    render(<Leaderboard games={GAME_DATA} />)
    const user = userEvent.setup();

    await user.click(screen.getByRole('button', { name: GAME_2_NAME }));

    expect(screen.getByRole('heading', { name: GAME_2_PLAYER})).toBeInTheDocument();
  })
});
