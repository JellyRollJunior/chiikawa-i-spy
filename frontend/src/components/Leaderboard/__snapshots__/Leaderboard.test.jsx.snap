// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Leaderboard component > renders title, game buttons, and winners 1`] = `
<div>
  <section
    class="_marginTopMedium_5890fc"
  >
    <h2
      class="_card_5890fc _title_5890fc"
    >
      Leaderboard
    </h2>
    <ul
      class="_leaderboard_acb05c _card_5890fc _marginTopSmall_5890fc"
    >
      <div
        class="_leaderboardButtonWrapper_acb05c"
      >
        <button
          class="_primaryButton_5890fc"
        >
          game 1
        </button>
        <button
          class="_primaryButton_5890fc"
        >
          game 2
        </button>
      </div>
      <li
        class="_leaderboardLegend_acb05c"
      >
        <h3>
          Place
        </h3>
        <h3>
          Name
        </h3>
        <h3>
          Time
        </h3>
      </li>
      <li
        class="_leaderboardItem_acb05c"
      >
        <h4>
          0
        </h4>
        <h4>
          Anonymous mob character
        </h4>
        <h4
          class="_left_acb05c"
        >
          7.383
          s
        </h4>
      </li>
    </ul>
  </section>
</div>
`;
